posts/2020/crud-pattern-in-react-implementation.md

```diff
@@ -0,0 +1,130 @@
+
+# Overview
+
+In the <a href="/blog/crud-pattern-in-react-introduction">previous post</a>, we checked the structure of `Books` component that implements CRUD actions and passes these actions and data (retrieved from API) to its child components `BookTable` and `BookForm`. In this post, We will check implementation details of `BookTable` and `BookForm` components.
+
+# BookTable Component - Implementation
+
+A table view is generally used to provide summary of the data. It may also provide an interface (eg. button) to trigger _Update_ and _Delete_ actions on the selected data record. The `BookTable` component presents book entities in a summary view in table format. It triggers _Update_ and _Delete_ actions using `showEditForm` and `performDelete` callback functions, passed as `props` by the `Books` component (parent). It also passes the data required for the actions (book object, book id) back to the parent component.
+
+React supports Unidirectional/One-Way Data Flow only. It means, the data is passed from parent to child as `props` which are immutable arguments. Our actions (update, delete) are defined in the `Books` component (parent), but we need the data (selected book id or object) from `BookTable` component (child) to perform the actions. We can achieve this by explicitly mapping Inverse Data Flow from `BookTable` component to `Books` component.
+
+## Inverse Data Flow
+
+- A callback function (`openForm()` or `delete()`) defined in the parent (`Books`) component is passed as props to the child (`BookTable`) component.
+- The function is invoked by the child component for an event .
+- eg. to trigger update or delete action (function invoke) on button click (event).
+- The child component can pass data (book id or book object) as an argument to the callback function. As the callback function is defined in the parent, the data is made available to the parent for further action.
+- As data is passed in the _inverse_ direction (from child to parent), Inverse Data Flow is achieved.
+- The passed data can then be used by the parent component to update internal state which may potentially re-render the page.
+
+# BookForm Component - Implementation
+
+A form component creates or updates an entity on the server by performing API actions. In order to do this, it needs to fulfill following requirements:
+
+- Validate form and display validation errors (if any) of specific field
+- Display API response errors (if any) of specific field on the UI
+
+We can fulfill these requirements by using a `formFieldAttributes` object that defines validation rules for each field. We will check its details in the next section.
+
+## `formFieldAttributes` object
+
+It defines custom validation rules for every field in `{fieldName: { attr1: value1, ..., attrN: valueN }}` format. The rules can be specified as attributes as following :
```

Comment:
`formFieldAttributes` is an object. This object is a collection of objects where each object is specific to a particular field in the form. These form specific fields wrap the logic required for validation and mapping the results to form fields. ~~It defines custom validation rules for every field in `{fieldName: { attr1: value1, ..., attrN: valueN }}` format. The rules can be specified as attributes as following :~~